cranelift/codegen/src/isa/aarch64/lower.rs

```diff
@@ -544,105 +544,242 @@ pub(crate) fn alu_inst_immshift(
-                // Extended addend.
-                let r2 = put_input_in_reg(
-                    ctx,
-                    InsnInput {
-                        insn: ext_insn,
-                        input: 0,
-                    },
-                    NarrowValueMode::None,
-                );
-                let r2_bits = ty_bits(ctx.input_ty(ext_insn, 0));
-                parts = Some((
-                    r1,
-                    r2,
-                    r2_bits,
-                    /* is_signed = */ op == Opcode::Sextend,
-                ));
-                break;
-            }
-        }
+    debug!(
+        "lower_address: addends64 {:?}, addends32 {:?}, offset {}",
+        addends64, addends32, offset
+    );
 
-        if let Some((r1, r2, r2_bits, is_signed)) = parts {
-            match (r2_bits, is_signed) {
-                (32, false) => {
-                    return MemArg::RegExtended(r1, r2, ExtendOp::UXTW);
-                }
-                (32, true) => {
-                    return MemArg::RegExtended(r1, r2, ExtendOp::SXTW);
-                }
-                _ => {}
-            }
+    // First, decide what the `MemArg` will be. Take one extendee and one 64-bit
+    // reg, or two 64-bit regs, or a 64-bit reg and a 32-bit reg with extension,
+    // or some other combination as appropriate.
+    let memarg = if addends32.len() > 0 && addends64.len() > 0 {
+        let (reg32, extendop) = addends32.pop().unwrap();
+        let reg64 = addends64.pop().unwrap();
+        MemArg::RegExtended(reg64, reg32, extendop)
+    } else if offset > 0 && offset < 0x1000 && addends64.len() > 0 {
+        let reg64 = addends64.pop().unwrap();
+        let off = offset;
+        offset = 0;
+        MemArg::RegOffset(reg64, off, elem_ty)
+    } else if addends64.len() >= 2 {
+        let reg1 = addends64.pop().unwrap();
+        let reg2 = addends64.pop().unwrap();
+        MemArg::RegReg(reg1, reg2)
+    } else if addends64.len() == 0 && addends32.len() > 0 {
+        let tmp = ctx.alloc_tmp(RegClass::I64, I64);
+        let (reg1, extendop) = addends32.pop().unwrap();
+        let signed = match extendop {
+            ExtendOp::SXTW => true,
+            ExtendOp::UXTW => false,
+            _ => unreachable!(),
+        };
+        ctx.emit(Inst::Extend {
+            rd: tmp,
+            rn: reg1,
+            signed,
+            from_bits: 32,
+            to_bits: 64,
+        });
+        if let Some((reg2, extendop)) = addends32.pop() {
+            MemArg::RegExtended(tmp.to_reg(), reg2, extendop)
+        } else {
+            MemArg::reg(tmp.to_reg())
         }
-    }
+    } else if addends64.len() == 0 && addends32.len() == 0 {
+        let off_reg = ctx.alloc_tmp(RegClass::I64, I64);
+        lower_constant_u64(ctx, off_reg, offset as u64);
+        offset = 0;
+        MemArg::reg(off_reg.to_reg())
+    } else {
+        let reg1 = addends64.pop().unwrap();
```

Comment:
From just skimming this code, it's hard to believe that this last branch is hit when `addends64.len() > 0`. Could the whole series be reorganized as `if addends64.len() > 0 { /* subchecks, etc. */ } else /*other subchecks*/{}`?

Reply:
Ah yes, it's much cleaner with that split -- thanks!